bpf: xdp generalization prep work

borkmann opened the pull request:
Last bigger chunk of changes needed for XDP that touches many places in our code base. After that, only local changes coming that don't require rebase anymore.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10491)
<!-- Reviewable:end -->

borkmann committed:
bpf: further prep-work to generalize contexts

- Abstract skb->cb[] usage into ctx_{load,store}_meta()
- Add ctx_get_ifindex() implementation for skb/xdp context
- Add ctx_adjust_meta() for use with ctx_{load,store}_meta()
- Add ctx_get_protocol() implementation for skb
- Remaining conversion of ctx_full_len()
- Remove unused helpers in bpf/ctx/skb.h and add stubs for xdp
  so that compiler doesn't optimize paths out (yet). Marked
  as BPF_STUB().
- Helper implementation for XDP in native code. Needs inline
  asm here as otherwise verifier is not able to track it.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: sync uapi header to upstream

Pull in the latest goodies from upstream.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: add bpf_mark_snat_done helper

Add a new helper bpf_mark_snat_done() which wraps setting the
MARK_MAGIC_SNAT_DONE flag. This only needs to be enabled when
we have __ctx_skb since we don't need to deal with egress packets
from XDP side.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: make recirculation marker generic

Rework the nodeport recirculation helpers into generic functions.

 - bpf_skip_nodeport_clear() -> initial clearing of NodePort recirc flag
 - bpf_skip_nodeport_set()   -> enable skip when we recirculate via tail call
 - bpf_skip_nodeport()       -> test whether we can skip

While skb side uses ctx->tc_index, xdp side implements it via packet
meta data.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: remove unneeded __overloadable attribute

With the current mechanism it's not needed anymore, so remove it from
our BPF code base.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

joestringer commented on bpf/cilium-map-migrate.c:
```diff
@@ -36,6 +36,7 @@
 #include "elf/libelf.h"
 #include "elf/gelf.h"
 
+#include "bpf/ctx/skb.h"
```
I didn't quite get why this userspace application needs this include?

borkmann replied:
To otherwise prevent a compile error since api.h either pulls in skb or xdp helpers. But I'll change it to unspec.h since we don't care about the context there.
```
In file included from include/bpf/helpers.h:9:0,
                 from include/bpf/api.h:14,
                 from cilium-map-migrate.c:40:
include/bpf/ctx/ctx.h:8:3: error: #error "No __ctx_buff context defined. Please either include 'bpf/ctx/skb.h' or 'bpf/ctx/xdp.h'."
 # error "No __ctx_buff context defined. Please either include 'bpf/ctx/skb.h' or 'bpf/ctx/xdp.h'."
   ^~~~~
```

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -20,6 +24,24 @@
 #endif
 
+static __always_inline __maybe_unused void
+bpf_skip_nodeport_clear(struct __ctx_buff *ctx)
```
I see, yeah this approach is what I was thinking we would probably end up doing to abstract away things like this. So from the core code it's not aware of the TC_INDEX but that's up to the implementation details of a particular BPF target hook. Abstraction is right, the trick is in getting all the details correct at XDP..

borkmann replied:
Yes, correct. So far this works nicely for what we need. Later if more code has to be pulled in for this layer, we would need to abstract more in a similar way.

joestringer commented on bpf/lib/l4.h:
```diff
@@ -37,7 +37,7 @@ static __always_inline int l4_modify_port(struct __ctx_buff *ctx, int l4_off,
 					  __be16 port, __be16 old_port)
 {
-	if (csum_l4_replace(ctx, l4_off, csum_off, old_port, port, sizeof(port)) < 0)
+	if (csum_l4_replace(ctx, l4_off, csum_off, old_port, port, 0) < 0)
```
Is this a bugfix?

Just surprised to see `sizeof(port)` passed into what seems to be a `flags` parameter.

borkmann replied:
No, it's not a bug fix. The current code works and the updated one, I've been converting it since these are the only few places left where the update-by-diff is not used. The other option would have been to bloat up l{3,4}_csum_replace() implementation for XDP with more cases. Once this overall work is complete, I'm planning to do benchmarking in general. I doubt these cases converted here make a difference overall.

joestringer reviewed (approved):
LGTM, minor nit below.

Too bad about the compiler not generating nice XDP code..

borkmann committed:
bpf: unbreak verifier after uapi header update

Syncing uapi header, I noticed that on older kernels (4.19.57 as one
example) the verification of sock_addr or __sk_buff programs fails due
to invalid context access. In particular instead of expected word-wise
access, compiler (here: clang-7/llvm-7) generates a double-word access
presumably due to changes in alignment. I'll investigate this in a
sepatate step. Note that when I was testing on a 5.x, the verifier was
passing just fine.

pahole dump:

  struct bpf_sock_addr {
        __u32                      user_family;          /*     0     4 */
        __u32                      user_ip4;             /*     4     4 */
        __u32                      user_ip6[4];          /*     8    16 */
        __u32                      user_port;            /*    24     4 */
        __u32                      family;               /*    28     4 */
        __u32                      type;                 /*    32     4 */
        __u32                      protocol;             /*    36     4 */
        __u32                      msg_src_ip4;          /*    40     4 */
        __u32                      msg_src_ip6[4];       /*    44    16 */

        /* XXX 4 bytes hole, try to pack */

        /* --- cacheline 1 boundary (64 bytes) --- */
        union {
                struct bpf_sock *  sk;                   /*    64     8 */
        } __attribute__((__aligned__(8)));               /*    64     8 */
        union {
                struct bpf_sock *          sk;                   /*     0     8 */
        };

        /* size: 72, cachelines: 2, members: 10 */
        /* sum members: 68, holes: 1, sum holes: 4 */
        /* forced alignments: 1, forced holes: 1, sum forced holes: 4 */
        /* last cacheline: 8 bytes */
  };

Verifier error:

  [...]
  level=warning msg=" R0=map_value(id=0,off=0,ks=2,vs=8,imm=0) R1_w=inv6 R6=ctx(id=0,off=0,imm=0) R7=map_value(id=0,off=0,ks=2,vs=8,imm=0) R8=map_value(id=0,off=0,ks=12,vs=12,imm=0) R9=inv(id=0,umax_value=65535,var_off=(0x0; 0xffff)) R10=fp0,call_-1" subsys=datapath-loader
  level=warning msg="249: (69) r2 = *(u16 *)(r7 +4)" subsys=datapath-loader
  level=warning msg=" R0=map_value(id=0,off=0,ks=2,vs=8,imm=0) R1_w=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R6=ctx(id=0,off=0,imm=0) R7=map_value(id=0,off=0,ks=2,vs=8,imm=0) R8=map_value(id=0,off=0,ks=12,vs=12,imm=0) R9=inv(id=0,umax_value=65535,var_off=(0x0; 0xffff)) R10=fp0,call_-1" subsys=datapath-loader
  level=warning msg="250: (63) *(u32 *)(r6 +24) = r2" subsys=datapath-loader
  level=warning msg="251: (63) *(u32 *)(r6 +20) = r1" subsys=datapath-loader
  level=warning msg="252: (18) r1 = 0xffff0000" subsys=datapath-loader
  level=warning msg="254: (63) *(u32 *)(r6 +16) = r1" subsys=datapath-loader
  level=warning msg="255: (b7) r1 = 0" subsys=datapath-loader
  level=warning msg="256: (7b) *(u64 *)(r6 +8) = r1" subsys=datapath-loader
  level=warning msg="invalid bpf_context access off=8 size=8" subsys=datapath-loader

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

qmonnet commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -4,52 +4,250 @@
+	 */
+	asm volatile("r1 = *(u32 *)(%[ctx] +0)\n\t"
+		     "r2 = *(u32 *)(%[ctx] +4)\n\t"
+		     "if %[off] > 0xff goto +6\n\t"
```
Any reason why not to move this check at the top and save 2 insns if offset is incorrect?

borkmann replied:
I was thinking about it originally, but then given it's slow path and we expect this to be never hit in normal case, it doesn't really matter.

qmonnet commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -4,52 +4,250 @@
+		     "if r1 > r2 goto +2\n\t"
+		     "%[ret] = 0\n\t"
+		     "goto +1\n\t"
+		     "%[ret] = -22\n\t"
```
Do we have a way to pass `EINVAL` instead of 22?

borkmann replied:
Need to double check, worst case via stringify() or such. Can follow-up on that.

qmonnet replied:
Don't bother too much about it though :)

borkmann replied:
Fixed in a better way, thx.

qmonnet commented on bpf/lib/overloadable_xdp.h:
```diff
@@ -50,4 +50,32 @@ redirect_self(struct xdp_md *ctx)
+#ifdef ENABLE_NODEPORT
+	return ctx_load_meta(ctx, RECIRC_MARKER);
+#else
+        return true;
```
Nit: indent

qmonnet commented on bpf/include/linux/bpf.h:
```diff
@@ -3449,7 +3456,9 @@ struct bpf_sock_addr {
 	__u32 msg_src_ip6[4];	/* Allows 1,2,4,8-byte read and 4,8-byte write.
 				 * Stored in network byte order.
 				 */
+#if 0 /* TODO */
```
It's 300 lines apart from previous comment, maybe add “See comment in struct __sk_buff”?

borkmann replied:
Still checking how we can work-around it to fix the issue w/o these workarounds.

qmonnet reviewed (commented):
I found a couple of minor things to suggest, but looks great!

borkmann committed:
bpf, xdp: pass in constants into inline asm instead of hard coding

Keep the code simple and readable.

Suggested-by: Quentin Monnet <quentin@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -4,52 +4,257 @@
+
+	if (!ctx_no_room(data_meta + off + 1, data))
+		data_meta[off] = datum;
+	else {
```
Nit: remove the braces here (or add braces to the `if`) for consistency.
